### PR TITLE
Enable direct streaming when possible

### PR DESCRIPTION
**Pull request by philippe44**
I've updated my bridges to support HTTPS and I've proposed a patch for squeezelite. 

This change in RP allows direct HTTPS streaming when player supports it. It probably requires more scrutiny, so I'm submitting it for your review. 

Particularly, I don't understand why you chose to use the $url parameter in parseDirectHeaders instead of $song->streamUrl (I had to make that change b/c $url points to radioparadise:// when direct streaming happens). 
The overloaded "new" method is not invoked when the canDirectStream returns an url, which only happens if the player has the 'canHTTPS' attribute (because RP subclasses Slim::Protocol::Player::HTTPS), but I'm also wondering if I'm not missing something here. Is all what 'new' does necessary, except for setting the $sock's url to a valid HTTP target (which is done by canDirectStream if player supports HTTPS). It seems to me that bitrate & contentType are set by parseDirectHeaders anyway, so isn't that redundant?

### CODE REVIEW

File: ProtocolHandler.pm [hunk 1]
```diff
@@ -72,7 +72,14 @@ sub getFormatForURL {
 # }
 
 sub canSeek { 0 }
-sub canDirectStreamSong { 0 }
+
+# allow players that can handle HTTPS to decide
+sub canDirectStream { 
+	my $class = shift;
+	my $client = shift;
+	my $url = shift;
```
Comment (michaelherger) [inline ProtocolHandler.pm:80]: Isn't the third parameter the `$song` object? Other implementations look like this:

```
sub canDirectStreamSong {
	my ( $class, $client, $song ) = @_;
	
	# We need to check with the base class (HTTP) to see if we
	# are synced or if the user has set mp3StreamingMethod
	return $class->SUPER::canDirectStream($client, $song->streamUrl(), $class->getFormatForURL());
}
```

I'll give that a try right now.

File: ProtocolHandler.pm [hunk 2]
```diff
@@ -207,15 +214,16 @@ sub parseDirectHeaders {
 
 	my $song = $client->streamingSong();
 	$ct =~ s/(?:m4a|mp4)/aac/i;
+	$ct = Slim::Music::Info::mimeToType($ct);
```
Comment (michaelherger) [inline ProtocolHandler.pm:217]: What is this part about? Necessary? Tested? What problem does it solve?

File: ProtocolHandler.pm (hunk 1 above)
Comment (michaelherger) [reply ProtocolHandler.pm:80]: Your version actually fails with
```
[19-04-28 20:27:28.9086] Slim::Networking::IO::Select::__ANON__ (131) Error: Select task failed calling Slim::Networking::Slimproto::client_readable: Can't locate object method "streamUrl" via package "Slim::Schema::RemoteTrack" at /Users/mh/Dropbox/Squeezebox/git/server/Plugins/RadioParadise/ProtocolHandler.pm line 359.
; fh=IO::Socket::INET=GLOB(0x7f926b56e8e0)
```

Comment (michaelherger) [reply ProtocolHandler.pm:80]: Oh... that's what the other changes are for :-). Never mind my previous comment.

Comment (philippe44) [reply ProtocolHandler.pm:80]: There are two canDirectStream methods. On that applies to a player $client and that one has  $client, $url and $song and the other one applies to a protocol handler and has $class, $client and $url. Here is what I have in Squeezebox2.pm
```
sub canDirectStream {
	my $client = shift;
	my $url = shift;
	my $song = shift;

	my $handler = Slim::Player::ProtocolHandlers->handlerForURL($url);

	if ($song && $handler && $handler->can("canDirectStreamSong")) {
		return $handler->canDirectStreamSong($client, $song);
	} elsif ($handler && $handler->can("canDirectStream")) {
		return $handler->canDirectStream($client, $url);
	}

	return undef;
}
```

File: ProtocolHandler.pm (hunk 2 above)
Comment (philippe44) [reply ProtocolHandler.pm:217]: I've seen that the expected contentType result of parseDirectHeaders is a not an audio/xxx but just the xxx part and that is used as a key of the player's supported formats hash. See line 549-690 of Squeezebox2.pm. Without that, when returned audio/x-flac (for example, this is what RP gives), then playback fails as format is claimed to be not recognized by player. This is also what HTTP.pm does

**Pull request closed without merging**